### PR TITLE
Better Emacs build on Mac OS

### PR DESCRIPTION
**Pull request by v-dobrev**
When installing on Mac OS, build and install the native Emacs.app along with the standard executables.

Edit: also adds emacs v29.3.

### CODE REVIEW

**Review by alalazo — DISMISSED** [review]

**Comment by alalazo** [comment]
@spackbot maintainers

**Comment by alalazo** [comment]
Waiting for @alecbcs review before merging.

**Comment by alecbcs** [comment]
Hi @v-dobrev! Thanks for the PR! I'd argue this might be better as an optional variant of the package. I personally use emacs through the terminal and wouldn't really be interested in installing it as an additional Application within MacOS.

**Comment by v-dobrev** [comment]
> Hi @v-dobrev! Thanks for the PR! I'd argue this might be better as an optional variant of the package. I personally use emacs through the terminal and wouldn't really be interested in installing it as an additional Application within MacOS.

Interesting, I'd expect that most people would prefer the GUI version (that's what I use) -- I guess it's a matter of taste.

In any case, what should I call the variant to toggle the non-gui/gui build on Mac?

**Comment by alecbcs** [comment]
Actually I think calling the variant `gui` is a great idea! For Linux this could take the place of the X variant and then under MacOS you'd have the build generate the Application/emacs.app.

**Comment by v-dobrev** [comment]
@alecbcs, I added "gui" variant, however, I kept it separate from the "X" variant because I didn't want to break backward compatibility.

If you want to get rid of the "X" variant, please feel free to make any changes in this branch.

**Comment by v-dobrev** [comment]
ping: @alecbcs, @alalazo

**Review by alalazo — DISMISSED** [review]
Built both versions on:

* **Spack:** 0.21.0.dev0 (59d515c1b4dd3aac38e5307f7f31a8efa4580b93)
* **Python:** 3.9.6
* **Platform:** darwin-ventura-m1
* **Concretizer:** clingo

I'll leave to @alecbcs the final review and merge.

**Comment by v-dobrev** [comment]
ping: @alecbcs

**Comment by v-dobrev** [comment]
@alecbcs, can you please review this? Thanks!

**Comment by v-dobrev** [comment]
@alalazo, I'm not getting any response from @alecbcs. I'd like to proceed and get this merged -- should I ask someone else to take a look?

**Comment by alecbcs** [comment]
@v-dobrev apologies we've been busy with the recent release of `v0.22`. There isn't currently a precedent for adding MacOS style applications to Spack install prefixes so before this gets merged I'd like to talk to @tgamblin and the team to figure out how we want to approach this style of install.

File: var/spack/repos/builtin/packages/emacs/package.py
Comment (alecbcs) [inline var/spack/repos/builtin/packages/emacs/package.py]:
```suggestion
            if spec.satisfies("+gui"):
```

It's now recommended to use `spec.satisfies()` instead of `in`.

**Review by alecbcs — COMMENTED** [review]
@v-dobrev in principle this PR looks good to me, just one quick suggestion. I'll try to bring this up with the team within the next week so we can figure out how to proceed.

**Comment by v-dobrev** [comment]
I just added the latest emacs version, 29.3.

**Review by tgamblin — APPROVED** [review]

**Pull request merged**
